### PR TITLE
Update pion/webrtc to 2.0.27

### PR DESCRIPTION
**Pull request by albrow**
The latest release *might* fix the performance problems we were experiencing. Waiting to see what happens in CI in order to confirm.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -116,31 +116,47 @@ github.com/opentracing/opentracing-go v1.0.2 h1:3jA2P6O1F9UOrWVpwrIo17pu01KWvNWg
 github.com/opentracing/opentracing-go v1.0.2/go.mod h1:UkNAQd3GIcIGf0SeVgPpRdFStlNbqXla1AfSYxPUl2o=
 github.com/pion/datachannel v1.3.0 h1:gxt/xGufDn8Yylk0uJB231xbGQVlFjVps+KdUAUl5Ls=
 github.com/pion/datachannel v1.3.0/go.mod h1:lxFbZLIT+EBPmy5AiCv8M0CXkcuTL53A4cyagZiRrDo=
+github.com/pion/datachannel v1.4.1 h1:VwK+oAzq72vup9aXBxTeIYvVOgHj0MBvrJnzOgCvet4=
+github.com/pion/datachannel v1.4.1/go.mod h1:044W2uxb6Wnm1c7X/jjhztShfD4e9p9yUNtdhvrTtaI=
 github.com/pion/dtls v1.3.1 h1:FE9/ghyVbRX2YyZWC++l/zDXPHxuYGqyQb9tu8cYBMA=
 github.com/pion/dtls v1.3.1/go.mod h1:CjlPLfQdsTg3G4AEXjJp8FY5bRweBlxHrgoFrN+fQsk=
```
Comment (backkem) [inline go.sum]: Maybe run ´go mod tidy´ to clean up these 'old' dependencies.

**Comment by albrow** [comment]
Just updated dependencies to include the following unreleased changes:

- https://github.com/libp2p/go-libp2p-transport/pull/51
- https://github.com/pion/datachannel/pull/18
- https://github.com/pion/sctp/pull/44
- https://github.com/pion/webrtc/pull/653

All tests pass locally now. Just waiting to see what happens on CI.

**Comment by albrow** [comment]
@backkem all the tests pass if we disable the race detector. I think the sheer amount of goroutines we are spinning up is just too much for the race detector to handle, especially on the TravisCI machines, which are not particularly powerful.

I vote that we disable the race detector for the stress tests and only use it for the other tests. What do you think?

**Comment by backkem** [comment]
Ok, it would be nice if we could do this while also using the test suite directly again: https://github.com/libp2p/go-libp2p-webrtc-direct/blob/b213ddc5d17985af4a77a60757cdc5956d71e6b7/webrtcdirect_test.go#L28

**Comment by raulk** [comment]
Yes, indeed. The race detector gives up as soon as you spin up more than 8192 concurrent goroutines. It's easy to detect when we're running with the race checker via build tags, we tend to use https://github.com/ipfs/go-detect-race to avoid reinventing the wheel; the upcoming `libp2p/go-libp2p-testing` module will subsume this util.

The pattern is:

1. skip stress tests if we're running with the race detector.
2. run a CI build matrix with and without the race detector.

**Comment by backkem** [comment]
The update to webrtc v2.0.16 and go-libp2p-core seems to cause a deadlock of some sort. I don't have the time to look into this ATM :(

**Comment by backkem** [comment]
Figured out the problem currently being faced: a pion/datachannel.DataChannel is packet-based and go-mplex expects a stream based net.Conn which has slightly different characteristics (while having the same interface since they are both connection-oriented).

Overview of different types of connections as I see them:

| | Connection oriented<br>(net.Conn) | Connection-less<br>(net.PacketConn) |
| - | - | - |
| Stream based<br>(Read/write byte streams) | IP, all of libp2p for now  |  ? |
| Packet based<br>(Read/write bounded packets) | ICE, DataChannel  | UDP |

**Comment by backkem** [comment]
@albrow Last commit makes the entire test suite green. I opted to use a fixed buffer to overcome the stream vs packet based issue. This point definitely needs further investigation and/or discussion with the libp2p guys. However, I think it's mergable in the mean time. Do you mind reviewing?

File: conn.go [hunk 2]
```diff
@@ -356,17 +357,47 @@ func (c *Conn) Transport() tpt.Transport {
...
+	bufEnd   int
 }
 
 func (w *dcWrapper) Read(p []byte) (int, error) {
-	return w.channel.Read(p)
+	var err error
+
+	if w.bufEnd == 0 {
+		n := 0
+		n, err = w.channel.Read(w.buf)
```
Comment (albrow) [inline conn.go]: I must be missing something. We read from `w.buf` here but when do we ever write to it?

Comment (backkem) [reply conn.go]: This line means we read from the datachannel into the packet buffer.

The idea is to allow an upper layer to read the packet in chunks. This is allowed when reading from a stream but it causes you to lose data when reading on a packet based connection. I'm not sure if there can be better solutions without the upper layer taking into account that it's using a packet based connection. This also ties into a discussion of support for packet based transports in libp2p, e.g. UDP.

Comment (albrow) [reply conn.go]: Ah I was thrown off by the method name `Read`. But that line is actually "reading into" `w.buf` aka writing to it.

**Comment by albrow** [comment]
@backkem LGTM I think we should go ahead and merge this PR since it passes all the tests. We can separately start a broader discussion about how to handle stream vs. packet based connections.

(I can't leave a review in GitHub because I'm the original author of the PR. If you hit the approve button I think GitHub will let us merge).

**Pull request merged**